kf_lib_data_ingest/etl/extract/extract.py

```diff
@@ -376,3 +324,146 @@ def _chain_operations(self, df_in, operations):
                     index = col_index
         df_out.index = index
         return df_out
+
+    def _run(self):
+        """
+        :returns: A dictionary where a key is the URL to the extract_config
+            that produced the dataframe and a value is a tuple containing:
+            (<URL to source data file>, <extracted DataFrame>)
+        """
+        output = {}
+        for extract_config in self.extract_configs:
+            self.logger.info(
+                "Extract config: %s", extract_config.config_filepath
+            )
+            protocol, path = split_protocol(extract_config.source_data_url)
+            if protocol == 'file':
+                if path.startswith('.'):
+                    # relative paths from the extract config location
+                    path = os.path.normpath(
+                        os.path.join(
+                            os.path.dirname(extract_config.config_filepath),
+                            path
+                        )
+                    )
+                else:
+                    path = os.path.expanduser(path)
+
+            data_path = protocol + PROTOCOL_SEP + path
+
+            # read contents from file
+            try:
+                df_in = self._source_file_to_df(
+                    data_path,
+                    do_after_load=extract_config.do_after_load,
+                    load_func=extract_config.load_func,
+                    **(extract_config.loading_params)
+                )
+            except ConfigValidationError as e:
+                raise type(e)(
+                    f'In extract config {extract_config.config_filepath}'
+                    f' : {str(e)}'
+                )
+
+            self.logger.debug(
+                "Loaded DataFrame with dimensions %s", df_in.shape
+            )
+
+            # extraction
+            df_out = self._chain_operations(
+                df_in, extract_config.operations
+            )
+
+            # split value lists into separate rows
+            df_out = split_df_rows_on_splits(
+                                                df_out.reset_index()
+                                            ).set_index('index')
+            del df_out.index.name
+
+            # standardize values again after operations
+            df_out = self._clean_up_df(df_out)
+
+            output[extract_config.config_filepath] = (data_path, df_out)
+
+        # return dictionary of all dataframes keyed by extract config paths
+        return output
+
+    def _postrun_analysis(self, run_output):
```

Comment:
Can you add a line or two to the docstring that describes the analysis being done in this stage's postrun analysis? Just something that says we're verifying unique concept counts.

Reply:
👍